File: analisis_numerico/metodos/algoritmos/lagrange.py
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

def lagrange(x, y):
    """
    lagrange_interp: Interpolación de Lagrange con salida simbólica decimal.
    Entrada:
        x: nodos (valores de x)
        y: valores correspondientes
    Salida:
        pol: polinomio simbólico interpolante (decimal)
    """
    x = np.asarray(x)
    y = np.asarray(y)

    if x.ndim != 1 or y.ndim != 1:
        raise ValueError("Las entradas deben ser vectores.")
    if len(x) != len(y):
        raise ValueError("Los vectores x e y deben tener la misma longitud.")
    if len(np.unique(x)) < len(x):
        raise ValueError("Los valores de x deben ser distintos.")

    n = len(x)
    t = sp.Symbol('t')
    poly_sym = 0

    for i in range(n):
        Li = 1
        den = 1
        for j in range(n):
            if j != i:
                Li *= (t - sp.Float(x[j], 6))
                den *= (x[i] - x[j])
        Li_term = y[i] * Li / den
        poly_sym += Li_term

    poly_sym = sp.expand(poly_sym)
    poly_decimal = sp.N(poly_sym, 6)

    print("Polinomio de interpolación (forma simbólica, decimal):")
    print(poly_decimal)

    # Graficar
    f = sp.lambdify(t, poly_decimal, modules='numpy')
    t_vals = np.linspace(min(x)-1, max(x)+1, 500)
    plt.plot(t_vals, f(t_vals), label='Polinomio', linewidth=2)
    plt.plot(x, y, 'ro', label='Datos originales', markersize=8)
    plt.title('Polinomio de Interpolación de Lagrange')
    plt.xlabel('x')
    plt.ylabel('P(x)')
    plt.grid(True)
    plt.legend()
    plt.show()

    return poly_decimal

File: Lagrange.m
function [pol] = Lagrange(x, y)
    % Lagrange: Interpolación de Lagrange con manejo de errores.
    % Entrada:
    %   x: vector de valores distintos (abscisas)
    %   y: vector de valores correspondientes (ordenadas)
    % Salida:
    %   pol: coeficientes del polinomio interpolante

    % Validaciones básicas
    if ~isvector(x) || ~isvector(y)
        error('Las entradas deben ser vectores.');
    end

    if length(x) ~= length(y)
        error('Los vectores x e y deben tener la misma longitud.');
    end

    n = length(x);

    if length(unique(x)) < n
        error('Los valores de x deben ser distintos.');
    end

    % Inicialización
    Tabla = zeros(n, n);

    for i = 1:n
        Li = 1;
        den = 1;
        for j = 1:n
            if j ~= i
                paux = [1, -x(j)];
                Li = conv(Li, paux);
                den = den * (x(i) - x(j));
            end
        end
        Tabla(i, :) = y(i) * Li / den;
    end

    pol = sum(Tabla);  % Coeficientes del polinomio

    % Mostrar el polinomio simbólicamente en forma decimal (plana)
    syms t
    poly_sym = 0;
    grado = length(pol) - 1;
    for i = 1:length(pol)
        poly_sym = poly_sym + pol(i) * t^(grado - i + 1);
    end

    poly_sym = vpa(poly_sym, 6);  % Convertir a decimal con 6 cifras

    disp('Polinomio de interpolación (forma simbólica, decimal):')
    disp(char(poly_sym))

    % Graficar el polinomio
    fplot(poly_sym, [min(x)-1, max(x)+1], 'LineWidth', 2)
    hold on
    plot(x, y, 'ro', 'MarkerSize', 8, 'MarkerFaceColor', 'r')
    title('Polinomio de Interpolación de Lagrange')
    xlabel('x')
    ylabel('P(x)')
    grid on
    legend('Polinomio', 'Datos originales')
    hold off
end
